### DISCUSSION TITLE
where to get firmware for primehub and how to flash it

### DISCUSSION THREAD

**Discussion by SpudGunMan** [pybricks/support#588] [Q&A] [answered]
I might be daft here but I cant seem to find the newest firmware for spike/RI4 bricks its not included in the packages on https://github.com/pybricks/pybricks-micropython/packages/241855 that I can tell?

https://github.com/pybricks/support/issues/167 I understand its alpha but is there a condensed how-to  (this thread is over year old and I am not finding any valuable fruit to do the basics here)

I tried to clone and make Pybricks-Micropython - primehub. but the make file seems light weight in providing feedback and it has errored the readme is just the GitHub page but no details on compiling the firmware for testing?

I also cant figure out how to get the spike hub into FS mode or flashing mode is this needed? aside from the DFU mode on the prime brick what other mode for flashing or how do I access it?

I see notes saying this is well documented but I cant find any complete documentation for pybricksdev or how it operates beyond the readme.md, how is the command line connecting to the brick? it just sits at timeout or crashes out for no connections do I need to manually contact with the OS first?

**Comment by dlech** [comment, marked as answer]
> I cant seem to find the newest firmware for spike/RI4 bricks

It is under *Actions*. Go to https://github.com/pybricks/pybricks-micropython/actions/workflows/build.yml?query=is%3Asuccess and pick the latest build. Then scroll down to *Artifacts*.



> it has errored the readme is just the GitHub page but no details on compiling the firmware for testing?

Did you read <https://github.com/pybricks/pybricks-micropython/blob/master/CONTRIBUTING.md>?

> I also cant figure out how to get the spike hub into FS mode or flashing mode is this needed?

It depends on what you need to do. Currently `pybricksdev` can upgrade from certain LEGO firmware versions to Pybricks firmware using `pybricksdev flash firmware.zip` where `firmware.zip` is the firmware downloaded from GitHub Actions as mentioned above. The official LEGO firmware is what provides the "FS" mode. 

We haven't updated it for the lastest LEGO firmware, so you might have to install and older version of the official LEGO firmware from https://github.com/gpdaniels/spike-prime/tree/master/firmware. This can be done with `pybricksdev dfu restore xxx.bin`. This is where the hub has to be in DFU mode. This is done by powering off the hub and removing the USB cable. Then press and hold the Bluetooth button. While holding this button, plug in the USB cable. Keep holding the button until the Bluetooth light flashes red/green/blue.

> how is the command line connecting to the brick?

After the Pybricks firmware is installed, use https://beta.pybricks.com to connect to the hub and run programs.

**Comment by SpudGunMan** [comment]
thank you very much that helped greatly :)
